### PR TITLE
Support for anonymous classes

### PR DESCRIPTION
**Pull request by kukulich**

### CODE REVIEW

**Comment by kukulich** [comment]
I've have just pushed failing test. The PR have to be fixed before review.

File: src/Reflection/ReflectionClass.php [hunk 1]
```diff
@@ -242,7 +242,15 @@ public static function createFromNode(
...
     public function getShortName() : string
     {
-        return $this->node->name;
+        if (!$this->isAnonymous()) {
+            return $this->node->name;
+        }
+
+        $name = 'class@anonymous';
+        if ($this->getFileName() !== null) {
+            $name .= ' ' . $this->getFileName();
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: I would at least add the line number. Also, spaces are not really usable in class names, IIRC. The typical naming is something like `class@anonymous/in/6UEGQ0x7f5b24d0e025`

Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Return early instead of overwriting the variable

File: src/Reflection/ReflectionClass.php [hunk 2]
```diff
@@ -242,7 +242,15 @@ public static function createFromNode(
      */
     public function getShortName() : string
     {
-        return $this->node->name;
+        if (!$this->isAnonymous()) {
+            return $this->node->name;
+        }
+
+        $name = 'class@anonymous';
+        if ($this->getFileName() !== null) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Cache the file name in a variable - avoid calling `$this->getFileName()` multiple times.

Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Note: this isn't due to performance, but due to the fact that the PHP type system has no immutability/idempotent guarantees

File: src/Reflection/ReflectionClass.php [hunk 3]
```diff
@@ -746,6 +754,16 @@ public function getDocComment() : string
     }
 
     /**
+     * Is this an anonymous class?
+     *
+     * @return bool
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Docblock not needed

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Hmm, this doesn't seem consistent with PHP core, but I'm not sure (without checking `php-src` directly) what the rules are for how this should work...

Given code:

```php
<?php

$c = new class() {};

$r = new ReflectionClass($c);
var_dump($r->getShortName());
var_dump($r->isAnonymous());
```

Running from the file:

```sh
$ php test.php 
/home/james/Desktop/test.php:6:
string(58) "class@anonymous\000/home/james/Desktop/test.php0x7f89bab29019"
/home/james/Desktop/test.php:7:
bool(true)
```

Running from https://3v4l.org/vRHuk:

```
string(39) "class@anonymous/in/vRHuk0x7fb312a9a019"
bool(true)
```

File: src/Reflection/ReflectionObject.php [hunk 4]
```diff
@@ -86,8 +86,14 @@ public static function createFromInstance($object)
             throw new \InvalidArgumentException('Can only create from an instance of an object');
         }
 
+        $className = get_class($object);
+
+        if (strpos($className, 'class@anonymous') === 0) {
+            throw new \LogicException('Not implemented - Unable to locate anonymous class');
```
Comment (Ocramius) [inline src/Reflection/ReflectionObject.php]: Is this the same exception thrown by internals? Also, I think we can actually locate them, compared to internal reflection...

File: src/Reflection/ReflectionObject.php [hunk 5]
```diff
@@ -86,8 +86,14 @@ public static function createFromInstance($object)
             throw new \InvalidArgumentException('Can only create from an instance of an object');
         }
 
+        $className = get_class($object);
+
+        if (strpos($className, 'class@anonymous') === 0) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionObject.php]: This string should be an internal constant, defined somewhere in the component

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (asgrim) [inline src/Reflection/ReflectionClass.php]: Note: it may be that this is acceptable, because in real terms the class name is irrelevant anyway :)

File: src/Reflection/ReflectionObject.php (hunk 5 above)
Comment (asgrim) [inline src/Reflection/ReflectionObject.php]: Nice catch early on :+1:

File: src/Reflection/ReflectionClass.php (hunk 1 above)
Comment (kukulich) [inline src/Reflection/ReflectionClass.php]: The PHP source code is here: https://github.com/php/php-src/blob/master/Zend/zend_compile.c#L6247

Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Ah, so there's a `"\0"` in there :D

**Comment by Ocramius** [comment]
@kukulich given that #310 is merged, this patch requires a rebase. I attempted one locally, but am getting some failures:

```
Time: 21.49 seconds, Memory: 72.00MB

There were 3 errors:

1) Roave\BetterReflectionTest\NodeCompiler\CompilerContextTest::testGetFileName
Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure: AST failed to parse in located source (in /home/ocramius/Documents/roave/betterreflection/test/unit/NodeCompiler/CompilerContextTest.php)

/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Exception/ParseToAstFailure.php:26
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:83
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:55
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AbstractSourceLocator.php:50
/home/ocramius/Documents/roave/betterreflection/src/Reflector/ClassReflector.php:53
/home/ocramius/Documents/roave/betterreflection/test/unit/NodeCompiler/CompilerContextTest.php:47

Caused by
PHPUnit_Framework_Error_Notice: Undefined property: PhpParser\Node\Name::$name

/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:110
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:108
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:85
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:147
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:80
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:55
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AbstractSourceLocator.php:50
/home/ocramius/Documents/roave/betterreflection/src/Reflector/ClassReflector.php:53
/home/ocramius/Documents/roave/betterreflection/test/unit/NodeCompiler/CompilerContextTest.php:47

2) Roave\BetterReflectionTest\Reflection\ReflectionFunctionAbstractTest::testIsGenerator with data set #15 ('<?php function foo() { new Fo...oo); }', true)
Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure: AST failed to parse in located source (first 20 characters: <?php function foo())

/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Exception/ParseToAstFailure.php:26
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:85
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:55
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AbstractSourceLocator.php:50
/home/ocramius/Documents/roave/betterreflection/src/Reflector/FunctionReflector.php:34
/home/ocramius/Documents/roave/betterreflection/test/unit/Reflection/ReflectionFunctionAbstractTest.php:191

Caused by
TypeError: Argument 1 passed to Roave\BetterReflection\Identifier\IdentifierType::isMatchingReflector() must implement interface Roave\BetterReflection\Reflection\Reflection, null given, called in /home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php on line 113

/home/ocramius/Documents/roave/betterreflection/src/Identifier/IdentifierType.php:60
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:113
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:159
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:85
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:147
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:80
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:55
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AbstractSourceLocator.php:50
/home/ocramius/Documents/roave/betterreflection/src/Reflector/FunctionReflector.php:34
/home/ocramius/Documents/roave/betterreflection/test/unit/Reflection/ReflectionFunctionAbstractTest.php:191

3) Roave\BetterReflectionTest\Util\Autoload\ClassPrinter\PhpParserPrinterTest::testPrinting
Roave\BetterReflection\SourceLocator\Ast\Exception\ParseToAstFailure: AST failed to parse in located source (in /home/ocramius/Documents/roave/betterreflection/test/unit/Fixture/TestClassForPhpParserPrinterTest.php)

/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Exception/ParseToAstFailure.php:26
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:85
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:55
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AbstractSourceLocator.php:50
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AggregateSourceLocator.php:37
/home/ocramius/Documents/roave/betterreflection/src/Reflector/ClassReflector.php:53
/home/ocramius/Documents/roave/betterreflection/src/Reflection/ReflectionClass.php:180
/home/ocramius/Documents/roave/betterreflection/test/unit/Util/Autoload/ClassPrinter/PhpParserPrinterTest.php:32

Caused by
TypeError: Argument 1 passed to Roave\BetterReflection\Identifier\IdentifierType::isMatchingReflector() must implement interface Roave\BetterReflection\Reflection\Reflection, null given, called in /home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php on line 113

/home/ocramius/Documents/roave/betterreflection/src/Identifier/IdentifierType.php:60
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:113
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:108
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:85
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:147
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:80
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/Locator.php:55
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AbstractSourceLocator.php:50
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Type/AggregateSourceLocator.php:37
/home/ocramius/Documents/roave/betterreflection/src/Reflector/ClassReflector.php:53
/home/ocramius/Documents/roave/betterreflection/src/Reflection/ReflectionClass.php:180
/home/ocramius/Documents/roave/betterreflection/test/unit/Util/Autoload/ClassPrinter/PhpParserPrinterTest.php:32

--

There were 2 failures:

1) Roave\BetterReflectionTest\Reflection\ReflectionClassTest::testIsAnonymousWithAnonymousClassInNamespace
Failed asserting that true is false.

/home/ocramius/Documents/roave/betterreflection/test/unit/Reflection/ReflectionClassTest.php:502

2) Roave\BetterReflectionTest\Reflector\FindReflectionsInTreeTest::testAnonymousClassCreatedInFunction
Roave\BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection::__invoke(Mock_Reflector_43824a4f Object (...), PhpParser\Node\Stmt\Class_ Object (...), Roave\BetterReflection\SourceLocator\Located\LocatedSource Object (...), null): Roave\BetterReflection\Reflection\Reflection was not expected to be called more than 2 times.

/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:108
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:120
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:101
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:171
/home/ocramius/Documents/roave/betterreflection/vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php:85
/home/ocramius/Documents/roave/betterreflection/src/SourceLocator/Ast/FindReflectionsInTree.php:147
/home/ocramius/Documents/roave/betterreflection/test/unit/SourceLocator/Ast/FindReflectionsInTreeTest.php:179

ERRORS!
Tests: 1665, Assertions: 29365, Errors: 3, Failures: 2, Skipped: 7.
```

**Comment by kukulich** [comment]
@Ocramius I will improve this PR later today.

**Comment by kukulich** [comment]
Ok, the PR is ready for review :)

File: src/Reflection/ReflectionClass.php [hunk 6]
```diff
@@ -25,6 +25,8 @@
 
 class ReflectionClass implements Reflection, \Reflector
 {
+    public const ANONYMOUS_CLASS_NAME = 'class@anonymous';
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: I think it's a prefix, so `ANONYMOUS_CLASS_NAME_PREFIX`

File: src/Reflection/ReflectionClass.php [hunk 7]
```diff
@@ -236,7 +238,16 @@ public static function createFromNode(
      */
     public function getShortName() : string
     {
-        return $this->node->name;
+        if (!$this->isAnonymous()) {
+            return $this->node->name;
+        }
+
+        $fileName = $this->getFileName();
+        if ($fileName === null) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: This could cause collisions when evaluating multiple classes, no?

Even in code with no assigned file:

```php
<?php

eval('$a = new class() {};');
eval('$b = new class() {};');

var_dump(get_class($a), get_class($b));
```
```
string(58) "class@anonymous/in/ZB2DN(3) : eval()'d code0x7fb03707a07a"
string(58) "class@anonymous/in/ZB2DN(4) : eval()'d code0x7fb03707a07a"
```
https://3v4l.org/ZB2DN

File: src/Reflection/ReflectionClass.php (hunk 3 above)
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Docblock indeed not needed

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        $anonymousClassNodesOnSameLine = array_values(array_filter($nodeVisitor->getAnonymousClasses(), function (Class_ $node) : bool {
+            return $node->getLine() === $this->coreClassReflection->getStartLine();
+        }));
+
+        if (count($anonymousClassNodesOnSameLine) === 0) {
+            return null;
+        }
+
+        if (count($anonymousClassNodesOnSameLine) > 1) {
+            throw new TwoAnonymousClassesOnSameLine(sprintf('Two anonymous classes on line %d in %s', $this->coreClassReflection->getStartLine(), $fileName));
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Please generate the exception in a named constructor of `TwoAnonymousClassesOnSameLine` - the default constructor is to be avoided

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+     * @var CoreReflectionClass
+     */
+    private $coreClassReflection;
+
+    /**
+     * @param object $anonymousClassObject
+     */
+    public function __construct($anonymousClassObject)
+    {
+        $this->coreClassReflection = new CoreReflectionClass($anonymousClassObject);
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Since we don't have an `object` hint yet (in 7.1), we should assert here

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+    private function getReflectionClass(Reflector $reflector, IdentifierType $identifierType) : ?ReflectionClass
+    {
+        if (!$identifierType->isClass()) {
+            return null;
+        }
+
+        $fileName = FileHelper::normalizeWindowsPath($this->coreClassReflection->getFileName());
+
+        $ast = (new ParserFactory())->create(ParserFactory::PREFER_PHP7)->parse(file_get_contents($fileName));
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Should we assert that the file exists?

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor($nodeVisitor);
+        $nodeTraverser->traverse($ast);
+
+        /** @var Class_[] $anonymousClassNodesOnSameLine */
+        $anonymousClassNodesOnSameLine = array_values(array_filter($nodeVisitor->getAnonymousClasses(), function (Class_ $node) : bool {
+            return $node->getLine() === $this->coreClassReflection->getStartLine();
+        }));
+
+        if (count($anonymousClassNodesOnSameLine) === 0) {
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: `if (! $anonymousClassNodesOnSameLine) {`

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+            return null;
+        }
+
+        if (count($anonymousClassNodesOnSameLine) > 1) {
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: `if (isset($anonymousClassNodesOnSameLine[1])) {`

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+                return $this->anonymousClasses;
+            }
+        };
+
+        $nodeTraverser = new NodeTraverser();
+        $nodeTraverser->addVisitor($nodeVisitor);
+        $nodeTraverser->traverse($ast);
+
+        /** @var Class_[] $anonymousClassNodesOnSameLine */
+        $anonymousClassNodesOnSameLine = array_values(array_filter($nodeVisitor->getAnonymousClasses(), function (Class_ $node) : bool {
```
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Can I suggest moving this into the anonmyous class? `getAnonymousClassesOnLine(int $lineNumber) : ?Class_` and let that API throw

File: test/unit/Fixture/AnonymousClassInNamespace.php [hunk 14]
```diff
@@ -0,0 +1,13 @@
+<?php
```
Comment (Ocramius) [inline test/unit/Fixture/AnonymousClassInNamespace.php]: A possible scenario missing is `NestedAnonymousClasses`:

```php
<?php
return new class () {
    function getWrapped($index) {
        if (0 === $index) {
            return new class() {};
        }

        return new class() {};
    }
}
```

This would verify that:

 * we analyze nested class declaration nodes (in this case, within an `if` block)
 * at runtime, we know the line where a reflection class was generated (depending on `$index` here)

File: test/unit/SourceLocator/Type/AnonymousClassObjectSourceLocatorTest.php [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        self::assertCount(0, $reflections);
+    }
+
+    public function testExceptionIfTwoAnonymousClassesOnSameLine() : void
+    {
+        self::expectException(TwoAnonymousClassesOnSameLine::class);
+        self::expectExceptionMessage(sprintf('Two anonymous classes on line 113 in %s', FileHelper::normalizeWindowsPath(__FILE__)));
+
+        $class1 = new class {}; $class2 = new class {};
+       (new AnonymousClassObjectSourceLocator($class1))->locateIdentifier(
```
Comment (Ocramius) [inline test/unit/SourceLocator/Type/AnonymousClassObjectSourceLocatorTest.php]: Wouldn't this test parse the test class itself? I think it's more efficient to split out this line:
```php
<?php
$class1 = new class {}; $class2 = new class {};
return [$class1, $class2];
```
Into a separate file, for ease of testing/parsing

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php (hunk 8 above)
Comment (Ocramius) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Yes: that means that there will be a test for the named constructor ;-)

**Comment by Ocramius** [comment]
@kukulich looking really slick, well done!

**Comment by Ocramius** [comment]
@kukulich since I'm sitting in a "light talk", I'm taking this over and applying the nitpicks myself :+1:

**Comment by kukulich** [comment]
@Ocramius Please no :) I have almost everything already done :) I just need to figure two thing.

**Comment by Ocramius** [comment]
@kukulich alright :+1:

File: src/Reflection/ReflectionClass.php (hunk 7 above)
Comment (kukulich) [inline src/Reflection/ReflectionClass.php]: The `$fileName` can be empty only if you use `StringSourceLocator`. I hope I've solved it.

I throw an exception for evaled anonymous class currenly because it's little complicated to solve it. The PhpParser returns this:

```
array(1) {
  [0]=>
  object(PhpParser\Node\Expr\Eval_)#32 (2) {
    ["expr"]=>
    object(PhpParser\Node\Scalar\String_)#31 (2) {
      ["value"]=>
      string(20) "$a = new class() {};"
      ["attributes":protected]=>
      array(3) {
        ["startLine"]=>
        int(3)
        ["endLine"]=>
        int(3)
        ["kind"]=>
        int(1)
      }
    }
    ["attributes":protected]=>
    array(2) {
      ["startLine"]=>
      int(3)
      ["endLine"]=>
      int(3)
    }
  }
}
```
 
It can be probably solved but I have to figure out how to solve start and end line numbers. I think it would be better to do it in another PR.

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php (hunk 9 above)
Comment (kukulich) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Done.

File: src/Reflection/ReflectionClass.php (hunk 7 above)
Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: > It can be probably solved but I have to figure out how to solve start and end line numbers. I think it would be better to do it in another PR.

Hardcoding them would be OK, as well as using the file contents hash as the file name.

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php (hunk 10 above)
Comment (kukulich) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: I added commit with `FileChecker`. I've placed it to `SourceLocator` namespace because the `InvalidFileLocation` is also there. However it may be better to place the class to `Util` namespace or simple add `checkFile` method to already existing `FileHelper`.

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php (hunk 13 above)
Comment (kukulich) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Done.

File: src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php (hunk 8 above)
Comment (kukulich) [inline src/SourceLocator/Type/AnonymousClassObjectSourceLocator.php]: Done :)

File: test/unit/Fixture/AnonymousClassInNamespace.php (hunk 14 above)
Comment (kukulich) [inline test/unit/Fixture/AnonymousClassInNamespace.php]: Added more tests.

File: src/Reflection/ReflectionClass.php (hunk 7 above)
Comment (kukulich) [inline src/Reflection/ReflectionClass.php]: Yes but I need new NodeVisitor to find `Eval_` and its content. Then I need the parse the content and find ast of the anonymous class.

And there could be situations like two `Eval_` on same line, two anonymous classes in the `Eval_` content...

Comment (Ocramius) [inline src/Reflection/ReflectionClass.php]: Oh, I see what you mean. I think that goes to far for us anyway :-|

**Comment by kukulich** [comment]
Thank you :)
